tests/ui/unneeded_struct_pattern.rs

```diff
@@ -0,0 +1,154 @@
+    clippy::manual_unwrap_or_default,
+    clippy::manual_unwrap_or,
+    clippy::redundant_pattern_matching
+)]
+#![warn(clippy::unneeded_struct_pattern)]
+
+extern crate non_exhaustive_enum;
+use non_exhaustive_enum::*;
+
+fn main() {
+    match Some(114514) {
+        Some(v) => v,
+        None {} => 0,
+    };
+
+    match Some(1919810) {
+        Some(v) => v,
+        None { .. } => 0,
+    };
+
+    match Some(123456) {
+        Some(v) => v,
+        None => 0,
+    };
+
+    match Some(Some(123456)) {
+        Some(Some(v)) => v,
+        Some(None {}) => 0,
+        None {} => 0,
+    };
+
+    if let None {} = Some(0) {}
+    if let None { .. } = Some(0) {}
+    if let Some(None {}) = Some(Some(0)) {}
+    let None {} = Some(0) else { panic!() };
+    let None { .. } = Some(0) else { panic!() };
+    let Some(None {}) = Some(Some(0)) else { panic!() };
+
+    enum Custom {
+        HasFields {
+            field: i32,
+        },
+        HasBracketsNoFields {},
+        NoBrackets,
+        #[non_exhaustive]
+        NoBracketsNonExhaustive,
+        Init,
+    };
+
+    match Custom::Init {
+        Custom::HasFields { field: value } => value, // Should be ignored
+        Custom::HasBracketsNoFields {} => 0,         // Should be ignored
+        Custom::NoBrackets {} => 0,                  // Should be fixed
+        Custom::NoBracketsNonExhaustive {} => 0,     // Should be fixed
+        _ => 0,
+    };
+
+    match Custom::Init {
+        Custom::HasFields { field: value } => value, // Should be ignored
+        Custom::HasBracketsNoFields { .. } => 0,     // Should be ignored
+        Custom::NoBrackets { .. } => 0,              // Should be fixed
+        Custom::NoBracketsNonExhaustive { .. } => 0, // Should be fixed
+        _ => 0,
+    };
+
+    match Custom::Init {
+        Custom::NoBrackets {} if true => 0, // Should be fixed
+        _ => 0,
+    };
+
+    match Custom::Init {
+        Custom::NoBrackets {} | Custom::NoBracketsNonExhaustive {} => 0, // Should be fixed
+        _ => 0,
+    };
+
+    if let Custom::HasFields { field: value } = Custom::Init {} // Should be ignored
+    if let Custom::HasBracketsNoFields {} = Custom::Init {} // Should be ignored
+    if let Custom::HasBracketsNoFields { .. } = Custom::Init {} // Should be ignored
+    if let Custom::NoBrackets {} = Custom::Init {} // Should be fixed
+    if let Custom::NoBrackets { .. } = Custom::Init {} // Should be fixed
+    if let Custom::NoBrackets {} | Custom::NoBracketsNonExhaustive {} = Custom::Init {} // Should be fixed
+    if let Custom::NoBracketsNonExhaustive {} = Custom::Init {} // Should be fixed
+    if let Custom::NoBracketsNonExhaustive { .. } = Custom::Init {} // Should be fixed
+
+    // Should be ignored
+    let Custom::HasFields { field: value } = Custom::Init else {
+        panic!()
+    };
+    // Should be ignored
+    let Custom::HasBracketsNoFields {} = Custom::Init else {
+        panic!()
+    };
+    // Should be ignored
```

Comment:
nit: Note that these could be replaced with [error annotations](https://rustc-dev-guide.rust-lang.org/tests/ui.html#error-annotations) but both are used interchangeably in clippy atm

Reply:
Done.